EIPS/eip-3155.md

```diff
@@ -0,0 +1,241 @@
+Clients should be able to execute simple transactions as well as code and return traces. In the following, we will call this client CUT (client under test) and use go-ethereums `evm` binary for code examples. 
+
+### Inputs
+
+The CUT MUST be able to execute a statetest as described [here](https://www.ethdocs.org/en/latest/contracts-and-transactions/ethereum-tests/state_tests/index.html).
+Example :
+``` 
+evm --json --nomemory statetest foobar.json 
+```
+The CUT MUST spit out jsonl as defined below, and it MUST also report the final `stateRoot` hash after executing the test. The CUT SHOULD report whether the tests succeeded or failed.
+
+If the statetest contains a Fork identifier (e.g. `Berlin` or `Istanbul`) the CUT MUST run the statetest under the chain rules of this fork.
+
+The CUT SHOULD executed statetests if the filenames are written to STDIN. This allows to execute multiple tests on the same instance without having to set up and tear down the client.
+
+### Output
+
+The CUT MUST output a `json` object for EACH operation. The output SHOULD to be printed to STDERR.
+
+Required Fields: 
+- `pc` Program Counter
+- `op` OpCode
+- `gas` Gas left before executing this operation
+- `gasCost` Gas cost of this operation
+- `stack` Array of all values on the stack
+- `depth` Depth of the call stack
+- `returnStack` Array of values, Stack of the called function
+- `returnData` Data returned by function call
+- `refund` Amount of **global** gas refunded
+- `memSize` Size of memory array
+
+Optional Fields: 
+- `opName` Name of the operation
+- `error` Description of an error 
+- `memory` Array of all allocated values
+- `storage` Array of all stored values
+
+*Example:* 
+```
+{"pc":0,"op":96,"gas":"0x2540be400","gasCost":"0x3","memory":"0x","memSize":0,"stack":[],"depth":1,"error":null,"opName":"PUSH1"}
+```
+
+The `stack`, `memory` and `memSize` are the values _before_ execution of the op. 
+All array attributes (`stack`, `returnStack`, `memory`) MUST be initialized to empty arrays ("stack":[],) NOT to null.
+If `--nomemory` is passed to the CUT, the `memory` field is omitted (but the `memSize` field has to be printed).
+Clients SHOULD implement a way to disable recording the storage as the stateroot includes all storage updates anyway.
+
+The CUT MUST NOT output a line for the `STOP` operation if an error occured:
+*Example:* 
+```
+{"pc":2,"op":0,"gas":"0x2540be3fd","gasCost":"0x0","memory":"0x","memSize":0,"stack":["0x40"],"depth":1,"error":null,"opName":"STOP"}
+```
+
+### Summary and error handling
+
+At the end of execution, the CUT MUST print some summerical info, this info SHOULD have the following fields. The summary SHOULD be printed to STDOUT.
```

Comment:
Should this summary field be a single json object like each json trace line? I think it should be specified as such.